### PR TITLE
add setting insert_shard_id

### PR DESCRIPTION
**Pull request by ucasfl**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

Add setting `insert_shard_id` to support insert data into specific shard from distributed table.

Detailed description / Documentation draft:

In many cases, sharding key can't satisfy people's requirement to partition data to difference shards. In this case, they need partition data in advance, then insert into difference local tables respectively. With this settings, peopel can insert data into specify shard from distributed table and don't need to perceive local tables, just need to know how many shards the distributed table has. 

For example:
```sql
INSERT INTO dist_table .... settings insert_shard_id = 1
```
Will insert the data synchronously into the first shard of the distributed table.

### CODE REVIEW

**Comment by yiguolei** [comment]
Great feature !!!

File: src/Storages/Distributed/DistributedBlockOutputStream.h [hunk 1]
```diff
@@ -62,10 +62,10 @@ class DistributedBlockOutputStream : public IBlockOutputStream
 
     void writeSplitAsync(const Block & block);
 
-    void writeAsyncImpl(const Block & block, const size_t shard_id = 0);
+    void writeAsyncImpl(const Block & block, size_t shard_id = 0);
 
     /// Increments finished_writings_count after each repeat.
-    void writeToLocal(const Block & block, const size_t repeats);
```
Comment (azat) [inline src/Storages/Distributed/DistributedBlockOutputStream.h:68]: You forget to update writeToLocal in the module part (.cpp), but in general, what is purpose of this change? (unrelated changes may create unrelated conflicts)

Comment (ucasFL) [reply src/Storages/Distributed/DistributedBlockOutputStream.h:68]: `const` in here is unnecessary.

File: src/Core/Settings.h [hunk 2]
```diff
@@ -420,6 +420,7 @@ class IColumn;
     M(Bool, async_socket_for_remote, true, "Asynchronously read from socket executing remote query", 0) \
     \
     M(Bool, optimize_rewrite_sum_if_to_count_if, true, "Rewrite sumIf() and sum(if()) function countIf() function when logically equivalent", 0) \
+    M(UInt64, insert_shard_id, 0, "If non zero, when insert intoi a distributed table, the data will be inserted into the shard `insert_shard_id` synchronously. Possible values range from 1 to `shards_number` of corresponding distributed table", 0) \
```
Comment (AngleNet) [inline src/Core/Settings.h:423]: /s/intoi/into Misspelling ?

Comment (ucasFL) [reply src/Core/Settings.h]: Yes...

**Comment by alex-zaitsev** [comment]
It would be better to do it via virtual column instead of a setting. Currently, _shard_num virtual column is returned in SELECTs from distributed tables, so the natural extension is to use it in INSERTs in order to write to a specific shard.

**Review by Avogar — APPROVED** [review]

**Comment by Avogar** [comment]
>It would be better to do it via virtual column instead of a setting. Currently, _shard_num virtual column is returned in SELECTs from distributed tables, so the natural extension is to use it in INSERTs in order to write to a specific shard.

@alexey-milovidov What do you think?

**Comment by alexey-milovidov** [comment]
But we don't have support for INSERTing of virtual columns.

File: src/Common/ErrorCodes.cpp [hunk 3]
```diff
@@ -538,7 +538,8 @@
     M(999, KEEPER_EXCEPTION) \
     M(1000, POCO_EXCEPTION) \
     M(1001, STD_EXCEPTION) \
-    M(1002, UNKNOWN_EXCEPTION)
+    M(1002, UNKNOWN_EXCEPTION) \
+    M(1003, INVALID_SHARD_ID)
```
Comment (azat) [inline src/Common/ErrorCodes.cpp:542]: BTW it is better to use values < 999 for such exceptions (since error codes >= 999 is something specific)
